Regular method missing return type should not be assumed to be a static constructor

**Version Used**: 
 04252e from master, 21 May 19

**Steps to Reproduce**:
```c#
class X
{
    private static Foo(int i) {}
}
```
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATAAgA0BYAKAG8z9r8AHBASwDcJ59sBGANnwDEB7fgAoGAOxj4GASnzkAvmTlA===

**Expected Behavior**:
```
error CS1520: Method must have a return type
```

**Actual Behavior**:
```
error CS0515: 'X.X(int)': access modifiers are not allowed on static constructors
error CS1520: Method must have a return type
error CS0132: 'X.X(int)': a static constructor must be parameterless
```

In this case `Foo` is unlikely to be a failed attempt at a static constructor because (1) it has the wrong name (2) it has parameters (3) it has an access modifier. It is vastly more likely I just forgot the return type on a regular method.

If the method name does not match the type name, the compiler should not give diagnostics for static constructors.

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbol.cs
@@ -116,7 +116,10 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 
             if (methodKind == MethodKind.StaticConstructor)
             {
-                if ((mods & DeclarationModifiers.AccessibilityMask) != 0)
+                // Don't report ERR_StaticConstructorWithAccessModifiers if the ctor symbol name doesn't match the containing type name.
+                // This avoids extra unnecessary errors.
+                // There will already be a diagnostic saying Method must have a return type.
+                if ((mods & DeclarationModifiers.AccessibilityMask) != 0 && ContainingType.Name == this.Name)
                 {
                     diagnostics.Add(ErrorCode.ERR_StaticConstructorWithAccessModifiers, location, this);
                     mods = mods & ~DeclarationModifiers.AccessibilityMask;

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbol.cs
@@ -119,7 +119,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                 // Don't report ERR_StaticConstructorWithAccessModifiers if the ctor symbol name doesn't match the containing type name.
                 // This avoids extra unnecessary errors.
                 // There will already be a diagnostic saying Method must have a return type.
-                if ((mods & DeclarationModifiers.AccessibilityMask) != 0 && ContainingType.Name == this.Name)
+                if ((mods & DeclarationModifiers.AccessibilityMask) != 0 &&
+                    ContainingType.Name == ((ConstructorDeclarationSyntax)this.SyntaxNode).Identifier.ValueText)
                 {
                     diagnostics.Add(ErrorCode.ERR_StaticConstructorWithAccessModifiers, location, this);
                     mods = mods & ~DeclarationModifiers.AccessibilityMask;

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs
@@ -60,10 +60,11 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             _lazyReturnType = TypeWithAnnotations.Create(bodyBinder.GetSpecialType(SpecialType.System_Void, diagnostics, syntax));
 
             var location = this.Locations[0];
-            // Don't report ERR_StaticConstructorWithAccessModifiers if the ctor symbol name doesn't match the containing type name.
+            // Don't report ERR_StaticConstParam if the ctor symbol name doesn't match the containing type name.
             // This avoids extra unnecessary errors.
             // There will already be a diagnostic saying Method must have a return type.
-            if (MethodKind == MethodKind.StaticConstructor && (_lazyParameters.Length != 0) && ContainingSymbol.Name == this.Name)
+            if (MethodKind == MethodKind.StaticConstructor && (_lazyParameters.Length != 0) &&
+                ContainingType.Name == ((ConstructorDeclarationSyntax)this.SyntaxNode).Identifier.ValueText)
             {
                 diagnostics.Add(ErrorCode.ERR_StaticConstParam, location, this);
             }

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs
@@ -60,7 +60,10 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             _lazyReturnType = TypeWithAnnotations.Create(bodyBinder.GetSpecialType(SpecialType.System_Void, diagnostics, syntax));
 
             var location = this.Locations[0];
-            if (MethodKind == MethodKind.StaticConstructor && (_lazyParameters.Length != 0))
+            // Don't report ERR_StaticConstructorWithAccessModifiers if the ctor symbol name doesn't match the containing type name.
+            // This avoids extra unnecessary errors.
+            // There will already be a diagnostic saying Method must have a return type.
+            if (MethodKind == MethodKind.StaticConstructor && (_lazyParameters.Length != 0) && ContainingSymbol.Name == this.Name)
             {
                 diagnostics.Add(ErrorCode.ERR_StaticConstParam, location, this);
             }

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs
@@ -20772,5 +20772,18 @@ partial class C {
                 //     partial void M(in int i) {}  
                 Diagnostic(ErrorCode.ERR_PartialMethodMustHaveLatent, "M").WithArguments("C.M(in int)").WithLocation(4, 18));
         }
+
+        [Fact]
+        public void MethodWithNoReturnTypeShouldNotComplainAboutStaticCtor()
+        {
+            CreateCompilation(@"
+class X
+{
+    private static Y(int i) {}
+}").VerifyDiagnostics(
+                // (4,20): error CS1520: Method must have a return type
+                //     private static Y(int i) {}
+                Diagnostic(ErrorCode.ERR_MemberNeedsType, "Y").WithLocation(4, 20));
+        }
     }
 }

